File: backend/app/schemas/proposal_rating.py
"""Schemas for proposal ratings."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import ConfigDict, Field, field_validator

from app.schemas.common import BaseSchema


class ProposalRatingUpsert(BaseSchema):
    """Payload for rating create/update."""

    model_config = ConfigDict(extra="forbid")

    coverage_needs_score: int = Field(..., ge=1, le=5)
    quality_info_score: int = Field(..., ge=1, le=5)
    business_data_score: int = Field(..., ge=1, le=5)
    comment: str | None = Field(default=None, max_length=1000)

    @field_validator("comment")
    @classmethod
    def validate_comment_whitespace(cls, value: str | None) -> str | None:
        if value is None:
            return None
        if value != "" and value.strip() == "":
            raise ValueError("Comment cannot be whitespace-only")
        return value


class ProposalRatingRead(BaseSchema):
    coverage_needs_score: int
    quality_info_score: int
    business_data_score: int
    comment: str | None
    updated_at: datetime


class ProposalRatingEnvelope(BaseSchema):
    rating: ProposalRatingRead | None


class ProposalRatingCriteriaAvg(BaseSchema):
    coverage_needs_avg: float
    quality_info_avg: float
    business_data_avg: float


class ProposalRatingStatsRead(BaseSchema):
    visible: bool
    rating_count: int
    minimum_required_count: int = 3
    overall_avg: float | None
    criteria_avg: ProposalRatingCriteriaAvg | None


class AdminProposalRatingListItem(BaseSchema):
    proposal_id: UUID
    project_id: UUID
    rating_count: int
    overall_avg: float
    criteria_avg: ProposalRatingCriteriaAvg
    latest_rating_at: datetime
    comment_count: int


class AdminProposalRatingListResponse(BaseSchema):
    items: list[AdminProposalRatingListItem]
    limit: int
    offset: int
    total: int


class ProposalRatingDistribution(BaseSchema):
    one: int = Field(alias="1")
    two: int = Field(alias="2")
    three: int = Field(alias="3")
    four: int = Field(alias="4")
    five: int = Field(alias="5")


class ProposalRatingDistributions(BaseSchema):
    coverage_needs_score: ProposalRatingDistribution
    quality_info_score: ProposalRatingDistribution
    business_data_score: ProposalRatingDistribution


class AdminProposalRatingComment(BaseSchema):
    comment: str
    updated_at: datetime


class AdminProposalRatingDetailResponse(BaseSchema):
    proposal_id: UUID
    project_id: UUID
    rating_count: int
    overall_avg: float
    criteria_avg: ProposalRatingCriteriaAvg
    distributions: ProposalRatingDistributions
    comments: list[AdminProposalRatingComment]


ProposalRatingsAdminHasComments = Literal["true", "false", "any"]
ProposalRatingsAdminSort = Literal["highest", "lowest", "mostRated", "recentlyRated"]

File: frontend/lib/types/proposal-rating.ts
export interface ProposalRating {
	coverageNeedsScore: number;
	qualityInfoScore: number;
	businessDataScore: number;
	comment: string | null;
	updatedAt: string;
}

export interface ProposalRatingEnvelope {
	rating: ProposalRating | null;
}

export interface ProposalRatingCriteriaAvg {
	coverageNeedsAvg: number;
	qualityInfoAvg: number;
	businessDataAvg: number;
}

export interface ProposalRatingStats {
	visible: boolean;
	ratingCount: number;
	minimumRequiredCount: number;
	overallAvg: number | null;
	criteriaAvg: ProposalRatingCriteriaAvg | null;
}

export interface ProposalRatingUpsertPayload {
	coverageNeedsScore: number;
	qualityInfoScore: number;
	businessDataScore: number;
	comment?: string | null;
}

export type AdminProposalRatingsSort =
	| "highest"
	| "lowest"
	| "mostRated"
	| "recentlyRated";

export type AdminProposalRatingsHasComments = "true" | "false" | "any";

export interface AdminProposalRatingsListParams {
	minOverall?: number;
	hasComments?: AdminProposalRatingsHasComments;
	ratedFrom?: string;
	ratedTo?: string;
	sort?: AdminProposalRatingsSort;
	limit?: number;
	offset?: number;
}

export interface AdminProposalRatingsListItem {
	proposalId: string;
	projectId: string;
	ratingCount: number;
	overallAvg: number;
	criteriaAvg: ProposalRatingCriteriaAvg;
	latestRatingAt: string;
	commentCount: number;
}

export interface AdminProposalRatingsListResponse {
	items: AdminProposalRatingsListItem[];
	limit: number;
	offset: number;
	total: number;
}

export interface ProposalRatingDistribution {
	"1": number;
	"2": number;
	"3": number;
	"4": number;
	"5": number;
}

export interface AdminProposalRatingsDetailResponse {
	proposalId: string;
	projectId: string;
	ratingCount: number;
	overallAvg: number;
	criteriaAvg: ProposalRatingCriteriaAvg;
	distributions: {
		coverageNeedsScore: ProposalRatingDistribution;
		qualityInfoScore: ProposalRatingDistribution;
		businessDataScore: ProposalRatingDistribution;
	};
	comments: Array<{
		comment: string;
		updatedAt: string;
	}>;
}
